### PR TITLE
feat(core) embedded dns resolver

### PR DESCRIPTION
**Pull request by thefosk**
This PR drastically changes how DNS resolutions work.

* Removes the dnsmasq dependency by manually parsing the `/etc/hosts` and `/etc/resolv.d` files. During resolutions, the entries in `/etc/hosts` always have the highest priority, with fallback to the configured DNS server.
* It's compliant with the TTL value returned by the DNS resolution, caching the resolutions in the shared dict for the amount specified in the TTL field.
* The configuration property `dns_resolvers_available` has been removed, and `dns_resolver` has been introduced. By default the system uses the local settings configured at `/etc/resolv.d` (using the first `nameserver` and being compliant with any `options` specified, if any). The DNS resolver can be manually set by uncommenting `dns_resolver`.
* Both `A` and `SRV` resolutions are supported. `SRV` resolutions are enabled by default and can be disabled to gain some performance in scenarios when it's not required, setting the property `srv: false` into `dns_resolver`.
* Introduces `balancer_by_lua` in the configuration file.

We can wait `refactor/cli` to be merged into `next` before merging this, as you like.

The PR is almost complete:
* Missing a couple more tests
* Reviewing the code

### CODE REVIEW

File: kong/tools/dns_resolver.lua [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+function DnsResolver:resolve(address, port, enable_srv)
+  -- Retrieve from cache
+  local cache_key = cache.dns_key(address)
+  local result = cache.get(cache_key)
+  if result then
+    return {host = result.host, port = result.port}
+  end
+
+  -- Query for A record
```
Comment (Tieske) [inline kong/tools/dns_resolver.lua]: an SRV record is supposed to start with `_`, which is an illegal character for A records iirc. So instead of querying them both successively, check for the underscore and query either A or SRV.

File: kong/tools/dns_resolver.lua [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    if not err and #answers > 0 then -- Ignore the error because some DNS servers don't support SRV
+      local srv_answer = answers[1]
+      if srv_answer.port > 0 then
+        final_port = srv_answer.port
+      end
+    end
+  end
+
+  if a_answer.ttl > 0 then
+    cache.set(cache_key, {host = final_address, port = final_port}, a_answer.ttl)
```
Comment (Tieske) [inline kong/tools/dns_resolver.lua]: final_address is never set when using SRV records, only the port is.

File: kong/tools/dns_resolver.lua (hunk 1 above)
Comment (thefosk) [inline kong/tools/dns_resolver.lua]: > an SRV record is supposed to start with _

I am not aware of this convention. In all the Consul examples an SRV record looks like an A record. In my tests SRV records were just returning port information, not IP information, so querying both is required.

Any thoughts?

File: kong/tools/dns_resolver.lua (hunk 2 above)
Comment (thefosk) [inline kong/tools/dns_resolver.lua]: Yep, because SRV just return port information not IP information (in Consul). Maybe I am wrong, willing to debug it with me?

File: kong/tools/dns_resolver.lua (hunk 1 above)
Comment (Tieske) [inline kong/tools/dns_resolver.lua]: See [RFC2782](https://tools.ietf.org/html/rfc2782). But apparently consul has multiple ways of looking up the same service; 'Standard lookup' and 'RFC 2782 lookup', see https://www.consul.io/docs/agent/dns.html

The way I understand it if you define a service in Consul it constructs multiple ways to retrieve the same service info.

It will be known by this name; `[tag.]<service>.service[.datacenter].<domain>`
as well as; `_<service>._<protocol>.service[.datacenter][.domain]`

They return both a port as well as a hostname/ip. So in case of a hostname, a second lookup may be required.

try; `dig srvtest.thijsschreijer.nl SRV` (not Consul, but regular DNS)

File: kong/tools/dns_resolver.lua (hunk 2 above)
Comment (Tieske) [inline kong/tools/dns_resolver.lua]: my results;
````
MacBook-Pro:kong thijs$ dig srvtest.thijsschreijer.nl SRV

; <<>> DiG 9.8.3-P1 <<>> srvtest.thijsschreijer.nl SRV
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33090
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;srvtest.thijsschreijer.nl.	IN	SRV

;; ANSWER SECTION:
srvtest.thijsschreijer.nl. 599	IN	SRV	20 10 8000 10.0.0.44.
srvtest.thijsschreijer.nl. 599	IN	SRV	10 5 8001 www.thijsschreijer.nl.
srvtest.thijsschreijer.nl. 599	IN	SRV	10 5 8000 10.0.0.43.

;; Query time: 31 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Tue May 17 23:47:21 2016
;; MSG SIZE  rcvd: 142

MacBook-Pro:kong thijs$
````
The entries with the highest priorities should be used, in this case (priority 20, 10, 10), it means the 2 records with prio 10 (lower = higher priority). An then do a weighted round robin on that. Weights being 5 and 5.

So probably something like;

- if the name has no `_` do an A lookup
- if it fails, or there is an `_` do an SRV lookup
- if SRV returns a hostname, resolve that one

Because the default ttl = 0 with Consul, its going to be a hot path, so we shouldn't do A and SRV unless absolutely necessary.

**Comment by Tieske** [comment]
As a general comment; can we cache the dns resolver objects? this creates a new dns resolver object for every request (assuming ttl = 0). Can't we reuse them. Less setting up and tearing down, and less GC cycles.

File: kong/tools/dns_resolver.lua (hunk 2 above)
Comment (thefosk) [inline kong/tools/dns_resolver.lua]: So when trying with Consul on my local system, an SRV lookup looks like:

```
$ nslookup -port=8600 -type=SRV thomascook.service.consul 127.0.0.1
Server:		127.0.0.1
Address:	127.0.0.1#8600

thomascook.service.consul	service = 1 1 0 sinzone.local.node.dc1.consul.
thomascook.service.consul	service = 1 1 0 sinzone.local.node.dc1.consul.
```

And and A lookup looks like:

```
$ nslookup -port=8600 -type=A thomascook.service.consul 127.0.0.1
Server:		127.0.0.1
Address:	127.0.0.1#8600

Name:	thomascook.service.consul
Address: 52.202.204.53
Name:	thomascook.service.consul
Address: 54.88.102.252
```

In my case the `SRV` lookup doesn't return the IP resolution, just the ports.

**Comment by thefosk** [comment]
> can we cache the dns resolver objects?

Yes, we should - when trying to cache them though the system was returning errors after a few queries. In the resty-dns examples the resolver is initialized every time too: https://github.com/openresty/lua-resty-dns#synopsis

File: kong/tools/dns_resolver.lua (hunk 2 above)
Comment (thefosk) [inline kong/tools/dns_resolver.lua]: On a side note I would use `nslookup` instead of `dig`, because sometimes `dig` does some additional computation, while we want something that looks as close as possible to the raw UDP request.

Comment (Tieske) [inline kong/tools/dns_resolver.lua]: Not an IP, but it does return a hostname.

SRV record says; 2 identical entries for `thomascook.service.consul`, priority = 1, weight = 1, port = 0, host = sinzone.local.node.dc1.consul

So your target is `sinzone.local.node.dc1.consul:0`. Right?

Comment (Tieske) [inline kong/tools/dns_resolver.lua]: Consul reports (but not all DNS servers do) the resolved IP addresses in the additional section of the DNS response. Not yet supported by Resty Dns library, see [Resty resolver #10](https://github.com/openresty/lua-resty-dns/pull/10)

**Comment by Tieske** [comment]
> Yes, we should - but when trying to cache them though the system was returning errors after a few queries.

Did you provide a locking mechanism? you must make sure (due to the co-sockets nature) that every DNS resolver object only handles a single complete cycle at a time. Before reusing the object for the next query.

**Comment by thefosk** [comment]
That didn't really concern me a lot since the base example of resty-dns creates a new resolver anyways, which may be the intended use, or not have any particular performance impact. It's an open question.

Locking would slow down resolution across the workers on high throughputs, maybe more than initializing a new resolver. I will benchmark this branch to figure it out.

File: kong/tools/dns_resolver.lua (hunk 2 above)
Comment (thefosk) [inline kong/tools/dns_resolver.lua]: Having that PR ^ closed would be amazing, because we could get all the data in one request.

**Comment by Tieske** [comment]
Locking is rather simple. Something like this; Create two lists at module level; `free` and `busy`. Whenever you whenever a socket operation is done (presumably all through `resolve`) grab one from `free` and move it from `free` to `busy`. When the full resolution cycle is done (sockets are free again), before `resolve` returns, move it back from `busy` to `free`.

The problem is that when a request is sent, the coroutine/light-thread is yielded by openresty. Now if a second request comes in and uses the same sockets before the response is returned, it goes awry. Cosocket code is written as sync, but really is async. Yielding is implicit whenever you do a socket operation.

File: kong/tools/dns_resolver.lua (hunk 2 above)
Comment (Tieske) [inline kong/tools/dns_resolver.lua]: But be careful, the trustworthiness of the additional section of the dns response is considered very low. In a Consul scenario it should be ok, but imo it should be an option whether to trust that data or not in general.

**Comment by thefosk** [comment]
My point is that we may not need locking in the first place, because creating the resolver every time doesn't really have any performance slowdowns (which seems to be the [intended use](https://github.com/openresty/lua-resty-dns#synopsis) in the resty-dns examples I linked above).

At a closer look when the library is initialized it creates a new socket for every request (at https://github.com/openresty/lua-resty-dns/blob/master/lib/resty/dns/resolver.lua#L112), which is a non-blocking operation, and then sets a timeout on the socket (https://github.com/openresty/lua-resty-dns/blob/master/lib/resty/dns/resolver.lua#L133) in order to enforce the DNS resolution timeout.

Your system seems more like a pool of DNS resolvers which while may be doing the work, adds additional complexity. I would first benchmark the performance of the current system, and then take an action depending if any performance loss is obtained.

In other words, let's make it work, let's run it, and see :)

**Comment by sinzone** [comment]
we should test this with other Service discovery tools other than consul. Like zookeeper.

**Comment by thefosk** [comment]
I am aiming at supporting Consul and etcd.

**Comment by kiddkai** [comment]
Just updated my PR, hope that it got 🈴 soon~, then you guys can finishes this feature~~

**Comment by lagonnebula** [comment]
Great job, that is exactly what i needed :+1:  Hope this PR will be validate soon.

**Comment by lagonnebula** [comment]
Hey, i builded this branch to try it, but i have some problem with a service configuration. How do you set the query to do to the dns discovery service ?

File: kong/tools/dns_resolver.lua [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  if #answers <= 0 then
+    return nil, "could not find any record for "..address
+  end
+
+  local a_answer = answers[1]
+  local final_address = a_answer.address
+
+  -- Query for SRV record
+  local final_port = port
+  if enable_srv then
```
Comment (lagonnebula) [inline kong/tools/dns_resolver.lua]: It seems that on the enable_srv he didn't load corrrectly the configuration, I put the srv: true on the yaml configuration file but it is always nil

So when im querying consul, i never got the port of the service. I made it to work by adding a realy bad fix.
On resolver.lua
```lua
if dns_config.srv == nil then
     dns_config.srv = true
 end
```

before the call of the 
```lua 
dns:resolve(parsed_url.host, parsed_url.port, dns_config.srv)
```

**Comment by Tieske** [comment]
> we should test this with other Service discovery tools other than consul. Like zookeeper.

If we stick to the standards, they all should work. Specific optimization can be done later.

> My point is that we may not need locking in the first place, because creating the resolver every time doesn't really have any performance slowdowns (which seems to be the intended use in the resty-dns examples I linked above).

It's not so much in resolving where the slowdown is, but in the GC. Everything created also needs to be cleaned up. Anyway, currently I don't think they can be reused, see https://github.com/openresty/lua-resty-dns/issues/13

**Comment by Tieske** [comment]
So some more issues to deal with;

- A and AAAA records; see http://stackoverflow.com/questions/4082081/requesting-a-and-aaaa-records-in-single-dns-query so we need to do two queries
- CNAME records; try; `smtp.thijsschreijer.nl` (CNAME) which points to `thuis.thijsschreijer.nl` (CNAME) which points to `wdnaste.duckdns.org` (A) which resolves to an IP. This is also the order in which the results are returned. So in this case `answer[1].address == answer[2].address == nil` we would need `answer[3].address`

**Comment by hutchic** [comment]
https://github.com/docker/libkv kv abstraction library with support for consul, etcd and zookeeper. Might not fit the use case / technology stack but thought it worth mentioning

**Comment by bitsofinfo** [comment]
Whats the eta on this? We publish a microservice container endpoint info into consul via registrator. Will this PR cover the use-case where kong could leverage that named service info in consul to dynamically route all requests to the correct backends as containers come up/down?

**Comment by Tieske** [comment]
it will not make it in the upcoming 0.9. But is scheduled for the next one. Presumably 0.10 if all goes well 4 weeks after 0.9 release (as 2 weeks for an RC, 2 weeks later final).

**Comment by thefosk** [comment]
@bitsofinfo you so you know, you can currently use Consul if you are not leveraging `SRV` records but only `A` records.

**Comment by Tieske** [comment]
This branch is now superseded by #1541.

If anyone wants to test the current status, please see https://github.com/Mashape/kong/pull/1541#issuecomment-242846402
